File: src/medium/count_good_nodes_in_binary_tree.go
package medium

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func checkNode(root *TreeNode, x int) int {
	if root == nil {
		return 0
	}

	total := 0
	if root.Val >= x {
		total++
	}

	if root.Val > x {
		x = root.Val
	}

	return total + checkNode(root.Left, x) + checkNode(root.Right, x)
}

func goodNodes(root *TreeNode) int {
	return 1 + checkNode(root.Left, root.Val) + checkNode(root.Right, root.Val)
}

File: src/medium/count_good_nodes_in_binary_tree.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::rc::Rc;

struct Solution;

impl Solution {
    fn count_good_nodes(node: Option<Rc<RefCell<TreeNode>>>, max_val: i32) -> i32 {
        match node {
            Some(rc_node) => {
                let node_ref = rc_node.borrow();
                let val = node_ref.val;
                let new_max = max_val.max(val);

                (val >= max_val) as i32
                    + Solution::count_good_nodes(node_ref.left.clone(), new_max)
                    + Solution::count_good_nodes(node_ref.right.clone(), new_max)
            }
            None => 0,
        }
    }

    pub fn good_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(rc_root) = &root {
            let root_val = rc_root.borrow().val;
            Solution::count_good_nodes(root, root_val)
        } else {
            0
        }
    }
}
